packages/apollo-gateway/src/datasources/RemoteGraphQLDataSource.ts

```diff
@@ -183,9 +185,9 @@ export class RemoteGraphQLDataSource<TContext extends Record<string, any> = Reco
     >,
   ): ValueOrPromise<GraphQLResponse>;
 
-  public didEncounterError(error: Error, _request: Request) {
-    throw error;
-  }
+  public didEncounterError?(
+    requestContextWithError: { error: Error, response?: Response, context: TContext, request: Request }
```

Comment:
Ok, a lengthy comment I wrote got on this line was lost (possibly because of a bug between the keyboard and the chair) and I cannot recover it.  I'll try to come back to this later, but please forgive my brevity right now:

---

The object that we colloquially refer to as the "request context" is the [`GraphQLRequestContext`](https://github.com/apollographql/apollo-server/blob/a926b7eedbb87abab2ec70fb03d71743985cb18d/packages/apollo-server-types/src/index.ts#L95-L127).  It takes on a number of more narrowed forms further [down in the `apollo-server-types`](https://github.com/apollographql/apollo-server/blob/a926b7eedbb87abab2ec70fb03d71743985cb18d/packages/apollo-server-types/src/index.ts#L175-L218) file.  The _user context_, which is `TContext` in this signature proposal, is on the `context` property of that `GraphQLRequestContext<TContext>`.

The object you are proposing that we pass here to `didEncounterError` as `requestContextWithError` is *not* that "request context" (or even really a variation of it with merely `error` attached as the name suggests).  Specifically, the `request` and `response` (which are certainly what I think you intend on exposing!) is a `fetch.Request` and `fetch.Response`, respectively (though provided via our `apollo-server-env` package), whereas a `GraphQLRequestContext` would contain `{ request: GraphQLRequest, response: GraphQLResponse }`).

For now, I might just propose we merely add `response: http.Response` as the third parameter and not introduce a breaking change, barring further API design discussion.  If the need is to also have _user_ `context` (or it's important to have other `GraphQLRequestContext`-y things, like `logger`!) and it warrants a breaking change (which I will say, comes with the burden of making a more clear `CHANGELOG` record that explains the change to consumers), perhaps another approach could be to have the well-known `requestContext` as the first argument and the HTTP-transport specific properties as their own secondary parameter object, e.g.:

```js
    requestContext: GraphQLRequestContext<TContext>,
    http: { error: Error, response: Response, request: Request },`
```

Though will need to pause and consider whether we want the pattern to be "The request context is always the last argument", which can sometimes be more durable for other-use-cases. 

Anyhow, long-story-longer, but with forced brevity because I deleted everything: Maybe we can just add `response` as the third argument? (And probably rename both the positional parameters to `httpRequest` and `httpResponse` to reduce confusion?